### ISSUE TITLE
Drawings can have items with differing fills

### ISSUE THREAD

**Issue by mlissner** [pymupdf/PyMuPDF#1355]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)

I have a PDF here:

https://storage.courtlistener.com/recap/gov.uscourts.akb.25471/gov.uscourts.akb.25471.83.0.pdf

In the upper right it has a drawing that consists of:

 - a red square
 - with a smaller white square on top of it
 - with a smaller red square on top of it
 - with a smaller white square on top of it

Creating the effect of two concentric square borders. 

When I get this drawing via `get_drawings`, it comes back as a single drawing, and all of the draw commands within it have the same `fill` color, red. Indeed, the `items` key of the returned drawing returns a two-tuple of the draw command type, and a rect — no fill color for the individual draw commands at all. 

## To Reproduce (mandatory)

1. Look at the doc linked above. Note the concentric rectangles.
1. Run get_drawings() on the doc
2. Note that all four rectangles are in your response, but they all have a `fill` of `(1.0, 0.0, 0.0)` (red).

## Expected behavior (optional)

I'd expect the individual draw commands to have color associated with them.

## Your configuration (mandatory)
 - Operating system, potentially version and bitness

64 bit Ubuntu. Happening on AWS lambda too.

 - Python version, bitness

64 bit Python 3.8

 - PyMuPDF version, installation method (**wheel** or **generated** from source).

Um, it's a layer in AWS, version 1.19.1.

Thanks.

**Comment by JorjMcKie** [comment]
I was thinking about a solution for this type of weakness.
As explained, the information about a rectangle's area orientation is not contained in `page.get_drawings()`.
However, it certainly is detectable in the page's data.
There is extra code in PyMuPDF, which scans backwards through the identified lines and combines every three lines into one "re" item (if certain conditions are met). At this point, the rectangle's orientation would be detectable, but there is no way to communicate this fact - given the current API. Changes here would break others' application code.
Any idea?
We currently have `("re", Rect(x0, y0, x1, y1))` for a rectangle item. I could include an additional switch 1 or -1 depending on the area orientation, like `("re", Rect(x0, y0, x1, y1), -1)` for clockwise orientation.

**Comment by JorjMcKie** [comment]
Sorry to bug again, but I cannot stop thinking about a solution that retains the current API.
I think I know the best way out:
Already today, I can detect whether 4 lines in a row form a polygon, and I wll output a "quad" item `("qu", quad)` in this case.
The detection logic is not perfect yet and needs to be improved.
So in cases of a clockwise rectangle, I will simpy treat and return it as a quad!
Some little logic would then help detecting that the quad's area is orientend clockwise ...

**Comment by mlissner** [comment]
This is a tricky one. I had to do a bunch of reading to understand, since I've never heard of winding rules before. 

I'm not sure the right solution. I think it's a bit strange to make a rectangle a quad just because it's wound in the wrong direction. 

I don't think it's so bad to include a third item in the tuple for the winding direction. That seems more or less API compatible. 

But for either of these approaches, there's also a need to note the winding rule (non-zero or even-odd) for the drawing. I guess that's fairly easy to add to the drawing object.

----

But setting all that aside, I'm trying to remember that my goal is to identify rectangles that intersect with text, figure out if they're above or below the text, and then whether the text is visible or not.

If the winding direction of each path and the winding rule for the drawing were added to the API, I'd be baffled and sad when trying to figure out my goal. 

For my use case, it feels like the real solution is that the rectangles and the text *don't* intersect. They share the same x-y space, and the text looks like it's "inside" the rectangle, but due to the winding rules, the inside of the rectangle is actually outside of it.

**Comment by JorjMcKie** [comment]
> ... and the text looks like it's "inside" the rectangle, but due to the winding rules, the inside of the rectangle is actually outside of it.

I do agree with you! Who knows what motivations drove the guys at Adobe when putting down the PDF specifications decades ago. From a today's perspective, that mechanism of color flipflop appears overly complex.

I also think, their wording of "regarded being outside" causes misleading connotations: it really still is the **_inside_** of that innermost rectangle. It is more like that areas of paths under the NZW rule have an additional property called **_orientation_**, that behaves much like a sign: overlaps of the same number of opposing orientations will cancel the fill color.

Anyway, I am working on a failsafe detection of area orientation first - at least for reactangles. How this is best communicated through the API is a separate decision.

**Comment by JorjMcKie** [comment]
Faster than I expected. Here is how the output of your problem page would look like now:
```python
[{'closePath': False,  # that line-like black rectangle
  'color': None,
  'dashes': '[] 0',
  'even_odd': False,
  'fill': (0.0, 0.0, 0.0),
  'fill_opacity': 1.0,
  'items': [('re', Rect(78.0, 286.9200134277344, 324.0, 287.5199890136719), 1)],
  'lineCap': [0],
  'lineJoin': 0,
  'rect': Rect(78.0, 286.9200134277344, 324.0, 287.5199890136719),
  'seqno': 1,
  'stroke_opacity': 1,
  'type': 'f',
  'width': 0},
 {'closePath': False,  # the white rectangle
  'color': None,
  'dashes': '[] 0',
  'even_odd': False,
  'fill': (1.0, 1.0, 1.0),
  'fill_opacity': 1.0,
  'items': [('re',
             Rect(432.9599914550781, 86.8800048828125, 494.5199890136719, 139.91998291015625),
             1)],  # new: standard orientation (counter-clockwise)
  'lineCap': [0],
  'lineJoin': 0,
  'rect': Rect(432.9599914550781, 86.8800048828125, 494.5199890136719, 139.91998291015625),
  'seqno': 3,
  'stroke_opacity': 1,
  'type': 'f',
  'width': 0},
 {'closePath': False,  # all the red rectangles
  'color': None,
  'dashes': '[] 0',
  'even_odd': False,
  'fill': (1.0, 0.0, 0.0),
  'fill_opacity': 1.0,
  'items': [('re',
             Rect(430.67999267578125, 84.5999755859375, 496.79998779296875, 142.199951171875),
             -1),  # clockwise
            ('re',
             Rect(431.6400146484375, 85.44000244140625, 495.8399963378906, 141.239990234375),
             1),  # anti-clockwise
            ('re',
             Rect(432.4800109863281, 86.4000244140625, 495.0, 140.4000244140625),
             -1),  # clockwise
            ('re',
             Rect(435.239990234375, 89.03997802734375, 492.239990234375, 137.63995361328125),
             1)],  # anti-clowise
  'lineCap': [0],
  'lineJoin': 0,
  'rect': Rect(430.67999267578125, 84.5999755859375, 496.79998779296875, 142.199951171875),
  'seqno': 4,
  'stroke_opacity': 1,
  'type': 'f',
  'width': 0}]
```
I still need to do shakedown tests, but so far it looks good.

**Comment by mlissner** [comment]
I just noticed that there's an `even_odd` attribute on drawings that I didn't notice before. Now that I see that, this looks good to me as the API upgrade. 

The only thing I see is that this will only work for paths that do not self-intersect, I think. Take [this example from Wikipedia](https://en.wikipedia.org/wiki/File:Even-odd_and_non-zero_winding_fill_rules.png) (works better if you have dark mode disabled on Github):

![504px-Even-odd_and_non-zero_winding_fill_rules](https://user-images.githubusercontent.com/236970/139924114-4dc91e83-e21c-4b88-8673-553262246cf5.png)

In that shape, clockwise and anti-clockwise are from the perspective of each point, not an attribute that makes sense on the path itself. The path itself is sometimes clockwise and sometimes not. 

The PDF spec uses left/right instead of clockwise/anti-clockwise, which seems to make more sense to me.

In any case, so long as this only is an attribute of rectangles and other non-self-intersecting shapes, I guess it works.

**Comment by JorjMcKie** [comment]
Thanks for the interesting reference. It seems like the original PDF spec authors might not have been the inventors of that concept in the end.
Anyway - I don't like it. It makes a big difference between just conceptually talking about area orientation or winding numbers on the one hand and one the other hand having to practically / programmatically determine the background color of some given point.
In PyMuPDF terminology, a path dictionary - list item of `page.get_drawings()` - indeed has no such attribute as a winding number. Only the rule to apply is one of it attributes.
But the items of `path["items"]` jointly and / or separately may define intersecting areas, which then all have some some orientation because of the presence of key-value `path["even_odd"] = False`.
A figure like your reference can easily be created with PyMuPDF and also be forced to show the above behaviour. A first easy glimpse is part of the documentation [here](https://pymupdf.readthedocs.io/en/latest/shape.html), where the `Shape.finish()` method is explained.

You and I have been discussing PyMuPDF's current weakness in that it "forgets" the orientation of rectangle areas.
As a side note: information to determine this property in all other cases is **_currently contained_** in the output of `get_drawings()`. So in principle you can already today (except for the rectangles) determine the background color of any point located in any `path["rect"]`. However tedious this job certainly is ... look at this example made width PyMuPDF worth of 15 lines of code:
![grafik](https://user-images.githubusercontent.com/8290722/139961486-989e4ad8-f96b-4d95-8b96-7c1a7cd7e677.png)

**Comment by mlissner** [comment]
You didn't provide the 15 lines of code, but I assume that my task is to write some code to determine the background color for a point. To do so, I'll use the `even_odd` attribute of the drawing and the new winding direction attribute of the `items` list. I think that's going to be too difficult to get right, honestly. 

I'm thinking it might be easier to just clip the PDF to the rectangle, export it as a PNG, and then inspect the pixels of that png to see if they're all the same. Otherwise, I feel like I'm developing a PDF renderer, which seems like something I will fail to do correctly.

**Comment by JorjMcKie** [comment]
> Otherwise, I feel like I'm developing a PDF renderer, which seems like something I will fail to do correctly.

Exactly, I could not agree more!
And your use case is more less focussed on judging the situation of rectangluar areas ... which I hope I now support much better than before. Not only in terms of detecting the area spin, but also in making "re" entries out of 3 or 4 consecutive connected line - "l" - items in hopefully all applicable cases.

The code scnippet is this:
```python
import fitz

doc = fitz.open()
page = doc.new_page()
GRAY = (0.9, 0.9, 0.9)
RED = (1, 0, 0)
GOLD = (1, 1, 0)
BLUE = (1, 1, 0)

r1 = fitz.Rect(100, 100, 300, 300)
shape = page.new_shape()
shape.draw_rect(r1)
shape.finish(fill=GRAY, even_odd=False)
shape.draw_bezier(r1.tl, r1.tr, r1.bl, r1.br)
shape.draw_bezier(r1.br, r1.tr, r1.bl, r1.tl)
shape.draw_bezier(r1.bl, r1.tl, r1.br, r1.tr)
shape.draw_bezier(r1.tr, r1.tl, r1.br, r1.bl)
shape.draw_line(r1.bl, r1.tr)
shape.draw_line(r1.tr, r1.br)
shape.draw_line(r1.br, r1.bl)
shape.draw_line(r1.bl, r1.tr)
shape.draw_line(r1.tr, r1.tl)
shape.draw_line(r1.tl, r1.bl)
shape.finish(fill=GOLD, color=RED, closePath=False, even_odd=False)

shape.commit()
doc.save(__file__.replace(".py", ".pdf"))
```

**Comment by mlissner** [comment]
OK, I think I understand:

1. You're improving PyMuPDF to do a better job making rects out of line segments.

2. You're adding the spin direction to `re` objects in the `items` list.

3. It's up to me (or other users of PyMuPDF) to figure out what color various points are within the PDF.

Am I understanding correctly?

**Comment by JorjMcKie** [comment]
Yes, this is exactly what I wanted to say.

**Comment by JorjMcKie** [comment]
Resolved by version 1.19.2.
